### ISSUE TITLE
fork() does not clone the repetition penalty window (upstream, fixed at master)

### ISSUE THREAD

**Issue by lloyal-research** [lloyal-ai/liblloyal#38]
## Summary

`fork()` documents that it clones the sampler chain **including penalties**. At
our pin (`b9581`) it does not: a forked branch starts with an empty repetition
penalty window. It fails silently — the child keeps generating, just without the
penalty pressure the parent had accumulated.

Fixed upstream; we inherit the fix with the llama.cpp pin bump (#37).

## Root cause — upstream, one field

`struct llama_sampler_penalties` carries two pieces of state:

```cpp
ring_buffer<llama_token>              prev;
std::unordered_map<llama_token, int>  token_count;  // "frequency map to count token occurrences"
```

- `llama_sampler_penalties_apply` reads **only `token_count`**
  (`llama-sampler.cpp:2679` — `find`, miss, `continue`, no penalty applied)
- `llama_sampler_penalties_clone` copies **only `prev`** (`:2720`)

The clone is structurally intact and functionally inert: ring buffer populated,
frequency map empty.

**Other stateful sampler clones were checked and are complete** — `dry` copies
four state fields, `mirostat`/`mirostat_v2` copy `mu`+`rng`, `xtc` copies `rng`,
`logit_bias`/`top_n_sigma` are stateless. Penalties is the only one affected.

## It is `fork()`, not `retainOnly()`

The defect is in the **clone**, and `fork()` is the only thing that clones.
`retainOnly()` performs a `seq_keep` pass and releases loser slots; the winner's
`sampler_chain` handle is untouched, so its state is preserved. Verified by
assertion below.

A fork only loses something when the source branch has **generated** — `accept`
reaches a branch solely through `_storeCommit`. Forking a prefilled-but-never-
generated branch loses nothing, because the window is empty either way.

## Failure shape and bounds

Penalties reset at every fork boundary. Within a segment they work; across a
fork they do not, so a child can repeat the parent's pre-fork text unpenalised
until its own window fills. Bounded by `penalty_last_n` (default 64) — roughly
the first 64 tokens after a fork.

**Scope:**

| context | exposure |
|---|---|
| agent pool / research scaffold | **none** — the spine never generates, so fan-out forks inherit an empty window either way |
| any caller not enabling penalties | **none** — `create_chain` only adds the sampler when `penalty_repeat != 1.0 \|\| penalty_freq != 0.0 \|\| penalty_present != 0.0`; defaults are `1.0 / 0.0 / 0.0` |
| `diverge.ts`, warm `use-agent.ts` | real |
| **the README's documented search loop** | **real, and compounding** |

The last one is why this is filed separately. `README.md:310-335` shows:

```cpp
for (int turn = 0; turn < max_turns; turn++) {
    auto leaf = root.fork();       // turn >= 1: root has generated and accepted
    ... generate on leaf ...
    store.retainOnly(leaves[0].handle());
    root = std::move(leaves[0]);   // the winner becomes the next root
}
```

Turn 0 is clean (root prefilled, never accepted). From turn 1 the root **has**
generated, so every leaf forked from it drops the previous turn's history. The
best-of-N example in the same README is clean, for the same reason turn 0 is.

## Documentation that is currently false

Four places claim penalties are cloned. Exactly one term in each is wrong —
everything else in those lists does clone correctly:

| location | text |
|---|---|
| `README.md:113` | "sampler chain (**penalties**, PRNG, filters)" |
| `branch.hpp:1668` | "Sampler chain (**penalties**, PRNG, filters)" |
| `branch.hpp:13` | "Sampler chain (**penalties**, PRNG, top-k/p filters)" |
| `sampler.hpp:521` | `clone_chain` — "an independent copy of the chain **with the same state**" |

KV (`kv::seq_cp`), grammar, PRNG, metrics, logits snapshot, logit bias and
cached params all clone correctly — verified.

## Reproduction

Verified across **five model families** — Qwen3.5-4B, SmolLM2-1.7B,
Llama-3.2-1B, Phi-3.5-mini, Gemma-4-E2B — identical result in every one. The
cases set logits explicitly via `set_logits`, so the model's own distribution
never enters the assertion, which is why the result is model-independent.

| case | at `b9581` |
|---|---|
| grammar parser position survives the clone | PASS |
| PRNG state cloned, reseed breaks it | PASS |
| **penalty window survives the clone** | **FAIL** |
| **state survives promote-then-fork** | **FAIL** (`retainOnly` step PASSES) |

The two passing cases matter: they establish that `llama_sampler_clone` works in
general, isolating the fault to the penalties sampler specifically.

Tests are **held locally** pending the bump rather than landing a knowingly-red
suite in CI. Patch: `~/.claude/plans/fork-fidelity-tests.patch`, applied against
`tests/integration/branch_integration_test.cpp`.

<details>
<summary>Reproduction test (click to expand)</summary>

```cpp

namespace {

/// Greedy chain with repetition penalty and every filter disabled, so the
/// only thing between logits and the sampled token is the penalty window.
struct PenaltyOnlyParams {
  float temperature = 0.0f;      // greedy — no dist, deterministic
  int32_t top_k = 0;             // disabled
  float top_p = 1.0f;            // disabled
  float min_p = 0.0f;            // disabled
  float typical_p = 1.0f;        // disabled
  float penalty_repeat = 2.0f;   // active
  float penalty_freq = 0.0f;
  float penalty_present = 0.0f;
  int32_t penalty_last_n = 64;
  uint32_t seed = 7;
};

TEST_CASE("fork fidelity: the penalty window survives the clone") {
  REQUIRE_MODEL();
  LlamaBackendGuard guard;

  auto model = TestConfig::acquire_test_model();
  REQUIRE(model);
  llama_context_params cp = llama_context_default_params();
  cp.n_ctx = 512;
  cp.n_seq_max = 4;
  llama_context* ctx = llama_init_from_model(model.get(), cp);
  REQUIRE(ctx);

  BranchStore store(8);
  store.init_tenancy(ctx);
  PenaltyOnlyParams params;

  BranchHandle parent = create(ctx, model.get(), store, 0, params);
  REQUIRE(parent != INVALID_HANDLE);
  const int32_t n_vocab = get_n_vocab(parent, store);
  REQUIRE(n_vocab > 2);

  // Token 0 wins outright on raw logits; token 1 wins once 0 is penalised.
  // With penalty_repeat = 2.0 llama.cpp divides a positive logit, so an
  // accepted 0 drops 10 -> 5, below 1's 6.
  std::vector<float> logits(static_cast<size_t>(n_vocab), -60.0f);
  logits[0] = 10.0f;
  logits[1] = 6.0f;

  set_logits(parent, std::span<const float>(logits), store);
  for (int i = 0; i < 3; ++i) accept_token(parent, 0, store);

  BranchHandle child = fork(parent, store);
  REQUIRE(child != INVALID_HANDLE);
  set_logits(child, std::span<const float>(logits), store);

  // Control: same params, same logits, NO accept history.
  BranchHandle fresh = create(ctx, model.get(), store, 0, params);
  REQUIRE(fresh != INVALID_HANDLE);
  set_logits(fresh, std::span<const float>(logits), store);

  // Greedy sampling advances no PRNG and accepts nothing, so drawing from
  // all three is free of ordering effects.
  const llama_token from_fresh  = sample(fresh, store);
  const llama_token from_parent = sample(parent, store);
  const llama_token from_child  = sample(child, store);

  // Control: without a penalty window token 0 is the argmax. If this fails
  // the logit margin is wrong for this model and everything below is vacuous.
  REQUIRE_MESSAGE(from_fresh == 0,
                  "control: an unpenalised branch must pick the argmax");

  // Does the penalty work at all on the branch that accepted the tokens?
  // Separating this from the fork assertion is what tells us whether a
  // failure is about cloning or about penalties generally.
  REQUIRE_MESSAGE(from_parent == 1,
                  "precondition: the accepting branch must itself be penalised");

  CHECK_MESSAGE(from_child == from_parent,
                "forked child must inherit the parent's penalty window");

  prune(child, store);
  prune(fresh, store);
  prune(parent, store);
  store.drain();
  llama_free(ctx);
}

```

</details>

The decomposition is deliberate: `from_fresh` proves the setup discriminates at
all, `from_parent` proves penalties work on the branch that accepted, and only
then does `from_child` isolate the clone. Without the middle assertion a failure
could not be distinguished from "penalties are broken generally".

The promote-then-fork case adds one assertion nothing else covers — that
`retainOnly` preserves the promoted branch's own sampler state — so a future
regression there is attributable to promotion rather than cloning.

## Fix

The pin bump (#37). No code change needed on our side; upstream master already
copies `token_count`. When it lands, reapply the fidelity patch and confirm all
four cases pass — and confirm **2 of 4 fail before** the bump, since a gate that
was never red proves nothing.

Interim question: whether `README.md:113` warrants a note while the bump is
pending. Depends on whether anyone outside is running the documented iterative
search idiom today.

**Comment by lloyal-research** [comment]
## Confirmed unreachable in our own stack — traced, not inferred

The scope table above lists "any caller not enabling penalties" as unexposed.
That has now been traced end to end for the lloyal.node + HDK path, and the
conclusion is stronger than "no window to lose": **the penalties sampler is never
constructed at all.**

The chain:

1. `LloyalSamplingParams` declares every field as `std::optional<...>`,
   default-constructed **empty** (`lloyal-node/src/SessionContext.hpp:33-36`).
2. `adaptSamplingParamsFromJS` assigns a penalty field **only when the JS object
   carries it** (`SessionContext.cpp:56-67`) — an absent `penalties` group leaves
   all four optionals empty.
3. `sampler::create_chain` resolves them through
   `as_value(params.penalty_repeat, 1.0f)`, and `as_value` on an empty optional
   returns the default → `1.0 / 0.0 / 0.0`.
4. `create_chain` adds the penalties sampler only when
   `penalty_repeat != 1.0f || penalty_freq != 0.0f || penalty_present != 0.0f`.

So with no `penalties` in the JS params, step 4 is false and no penalties sampler
enters the chain. There is no state for `llama_sampler_clone` to drop.

**Nothing downstream sets them.** Grepping the HDK packages and the scaffold
templates for `penalties` / `penalty_repeat` returns exactly one hit, and it is a
type declaration: `packages/sdk/src/types.ts:676` — `penalties?: PenaltyParams`.
No assignment anywhere.

### What this changes

- **No runtime exposure on any lloyal path**, which is a stronger statement than
  the earlier "the spine never generates so fan-out forks lose nothing". Even
  `diverge` and warm `use-agent` — the two sites that *would* be exposed — cannot
  hit it, because they never construct the sampler either.
- The remaining exposure is **liblloyal as a published library**: a consumer who
  explicitly enables penalties AND forks from a branch that has generated. The
  documented search loop is that shape.
- The `README.md:113` correction therefore has **no deadline of its own** and can
  ride the pin bump, unless an external consumer is enabling penalties today —
  the one fact not visible from inside this repo.

### And what it means for the tests

The fidelity cases are not guarding a live path here; they are guarding a
**documented contract our own runtime does not currently exercise**. That is
still worth keeping — it is what surfaced an upstream defect that no header diff
or changelog scan could have found — but it argues for extending the suite to the
clone-carried state we *do* depend on (grammar parser position and PRNG across
fork, both already covered and both passing) rather than treating penalties as
the representative case.
